<!DOCTYPE html>
<html lang="en">

<head>
    <meta charset="utf-8">
    <title>Phylogenetic tree reconstruction based on evolutionary distance</title>
    <meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
    <meta name="author" content="">
    <link rel="shortcut icon" type="image/x-icon" href="/BIO-BTE-10-L-4//img/favicon.ico">
    <script src="https://code.jquery.com/jquery-3.4.1.min.js"></script>
    <script type="text/javascript" src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <!-- <script type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.7/MathJax.js?config=TeX-MML-AM_CHTML"></script> -->

    <link type="text/css" rel="stylesheet" href="/BIO-BTE-10-L-4/content/fsdocs-default.css" />
    <link type="text/css" rel="stylesheet" href="/BIO-BTE-10-L-4/content/fsdocs-custom.css" />
    <script type="text/javascript" src="/BIO-BTE-10-L-4/content/fsdocs-tips.js"></script>
    <script type="text/javascript" src="/BIO-BTE-10-L-4/content/init.js"></script>
    <script type="text/javascript" src="https://cdn.jsdelivr.net/npm/@creativebulma/bulma-collapsible@1.0.4/dist/js/bulma-collapsible.min.js"></script>
    <link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/@creativebulma/bulma-collapsible@1.0.4/dist/css/bulma-collapsible.min.css">
    <!-- HTML5 shim, for IE6-8 support of HTML5 elements -->
    <!--[if lt IE 9]>
      <script src="http://html5shim.googlecode.com/svn/trunk/html5.js"></script>
    <![endif]-->
    <!-- BEGIN SEARCH BOX: this adds support for the search box -->
    <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/JavaScript-autoComplete/1.0.4/auto-complete.css" />
    <!-- END SEARCH BOX: this adds support for the search box -->
    
</head> 

<body>
    <div class="columns is-fullheight m-0">
        <div id="fsdocs-menu" class="column is-2 is-paddingless">
            <aside class="menu">
                <div class="container is-paddingless is-marginless has-background-dark">
                    <a href="/BIO-BTE-10-L-4/">
                        <figure class="image is-128x128 container">
                            <img src="/BIO-BTE-10-L-4/img/logo.png">
                        </figure>
                    </a>
                    <h2 class="title has-text-centered has-text-white pb-4">BIO-BTE-10-L-4</h2>
                </div>
                <div class="content">
                    <!-- BEGIN SEARCH BOX: this adds support for the search box -->
                    <div id="header">
                        <div class="field">
                            <div class="control has-icons-left has-icons-right searchbox" id="fsdocs-searchbox">
                                <input data-search-input="" id="search-by" type="search" placeholder="Search..." class="input"/>
                                <span class="icon is-small is-left">
                                    <i class="fas fa-search"></i>
                                </span>
                            </div>
                        </div>
                    </div>
                    <!-- END SEARCH BOX: this adds support for the search box -->
                    <ul class="navbar-nav">
                        <li class="nav-header">
  general
</li>             
<li class="nav-item">
  <a class="nav-link" href="/BIO-BTE-10-L-4/general/blog-post-guide.html">
    Blog post guide
  </a>
</li>             
<li class="nav-item">
  <a class="nav-link" href="/BIO-BTE-10-L-4/general/markdown_cheatsheet.html">
    Markdown cheatsheet
  </a>
</li>             
<li class="nav-item">
  <a class="nav-link" href="/BIO-BTE-10-L-4/general/example-blog-post.html">
    Example blog post
  </a>
</li>             
<li class="nav-header">
  projects
</li>             
<li class="nav-item">
  <a class="nav-link" href="/BIO-BTE-10-L-4/projects/tSNE.html">
    t-Distributed Stochastic Neighbour Embedding
  </a>
</li>             
<li class="nav-item">
  <a class="nav-link" href="/BIO-BTE-10-L-4/projects/hClust.html">
    Efficient agglomerative hierarchical clustering
  </a>
</li>             
<li class="nav-item">
  <a class="nav-link" href="/BIO-BTE-10-L-4/projects/evolutionary-distance.html">
    Phylogenetic tree reconstruction based on evolutionary distance
  </a>
</li>             
<li class="nav-item">
  <a class="nav-link" href="/BIO-BTE-10-L-4/projects/spectral-alignment.html">
    Alignment of Peptide derived MS Spectra
  </a>
</li>             
<li class="nav-item">
  <a class="nav-link" href="/BIO-BTE-10-L-4/projects/PrimsAlgorithm.html">
    Reduce complex graphs to the best paths
  </a>
</li>             
<li class="nav-item">
  <a class="nav-link" href="/BIO-BTE-10-L-4/projects/sequencingByHybridization.html">
    Sequencing by Hybridization as an Eulerian Path Problem
  </a>
</li>             
<li class="nav-header">
  bachelor thesis
</li>             
<li class="nav-item">
  <a class="nav-link" href="/BIO-BTE-10-L-4/bachelor/molecular-evolution.html">
    Molecular evolution of thermolabile proteins in microalgae
  </a>
</li>
                    </ul>
                </div>
            </aside>
        </div>
        <div class="column is-10 is-paddingless">
            <div class="section pt-4">
                <div class="container" id="fsdocs-content">
                    
<h1><a name="Phylogenetic-tree-reconstruction-based-on-evolutionary-distance" class="anchor" href="#Phylogenetic-tree-reconstruction-based-on-evolutionary-distance">Phylogenetic tree reconstruction based on evolutionary distance</a></h1>
<p><strong>Interested?</strong> Contact <a href="mailto:muehlhaus@bio.uni-kl.de">muehlhaus@bio.uni-kl.de</a> or <a href="mailto:schneike@bio.uni-kl.de">schneike@bio.uni-kl.de</a></p>
<h4><a name="Table-of-contents" class="anchor" href="#Table-of-contents">Table of contents</a></h4>
<ul>
<li>
<a href="#Introduction">Introduction</a>
<ul>
<li><a href="#Phylogenetic-trees">Phylogenetic trees</a></li>
<li>
<a href="#Evolutionary-Distance-of-DNA-sequences">Evolutionary Distance of DNA sequences</a>
<ul>
<li><a href="#Proportional-distance">Proportional distance</a></li>
</ul>
</li>
<li>
<a href="#Distance-Corrections-based-on-evolutionary-Models">Distance Corrections based on evolutionary Models</a>
<ul>
<li><a href="#The-Jukes-Cantor-Model">The Jukes-Cantor Model</a></li>
<li><a href="#The-Kimura-2-Parameter-Model">The Kimura 2-Parameter Model</a></li>
</ul>
</li>
</ul>
</li>
<li><a href="#Aim-for-this-project">Aim for this project</a></li>
<li>
<a href="#Coding-clues">Coding clues</a>
<ul>
<li><a href="#Before-you-start">Before you start</a></li>
<li><a href="#Scripting-environment-and-necessary-libraries">Scripting environment and necessary libraries</a></li>
<li><a href="#General-coding-advice">General coding advice</a></li>
</ul>
</li>
<li><a href="#References">References</a></li>
<li>
<a href="#Additional-information">Additional information</a>
<ul>
<li><a href="#Testing">Testing</a></li>
<li><a href="#Blog-post">Blog post</a></li>
</ul>
</li>
</ul>
<h2><a name="Introduction" class="anchor" href="#Introduction">Introduction</a></h2>
<h3><a name="Phylogenetic-trees" class="anchor" href="#Phylogenetic-trees">Phylogenetic trees</a></h3>
<p>Phylogenetic trees are diagrams that visualize inferred evolutionary relationships between a set of organisms. Consider this tree diagram:</p>
<pre class="fssnip highlighted"><code lang="fsharp">         <span class="id">____</span> <span class="id">A</span>
       <span class="id">_</span><span class="pn">|</span>
      <span class="pn">|</span> <span class="pn">|</span><span class="id">____</span> <span class="id">B</span>
<span class="id">______</span><span class="pn">|</span>
      <span class="pn">|</span><span class="id">______</span> <span class="id">C</span>
</code></pre>
<p>Ancestors are shown as nodes on the tree, while the leaves represent the respective organisms.
It tells you that A and B share a common ancestor, and that that ancestor shares a common ancestor with C.
In other words, A and B are closer related to each other, than each of them to C.</p>
<p>A and B form a <em>clade</em> together with their common ancestor (also known as a <em>monophyletic group</em>) - a group of organisms that includes a single ancestor
and all of its descendents that represent unbroken lines of evolutionary descent.</p>
<p>But based on what information do we construct such trees? There are different classes of approaches to this problem, but to stay beginner-friendy,
only <em>distance-based</em> methods will be discussed in the scope of this project. For sake of completeness, other approaches include <em>parsimony</em>,
<em>maximum likelihood</em> and <em>Bayesian approaches</em> to searching the possible tree space.</p>
<p>The first step in any (distance-based) phylogenetic tree reconstruction is the selection of the characteristic to infer evolutionary relationships from,
and subsequently the determination of the phylogenetic distance between the organisms of interest based on that characteristic.</p>
<br>
<hr />
<h3><a name="Evolutionary-Distance-of-DNA-sequences" class="anchor" href="#Evolutionary-Distance-of-DNA-sequences">Evolutionary Distance of DNA sequences</a></h3>
<p>Any kind of characteristic of organisms can be used to try to infer phylogenetic relationships - like for example beaks of Darwin finches -
but DNA sequences have proven to be incredibliy helpful to reconstruct phylogenetic trees, as the nucleotide alphabet is relatively
simple and sequencing data has reached unparalleled throughput and accuracy. Likewise, there are a wide range of sophisticated methods to
calculate phylogenetic distance based on DNA sequences.</p>
<p>In the scope of this project, you will take a step back and look at some classic evolutionary models that can be used to model phylogenetic
distance based on DNA sequences.</p>
<p>A few important bits of jargon for the following chapters:</p>
<ul>
<li>
DNA substitution mutations can be classified by 2 types:
<ul>
<li><em>Transitions</em> are interchanges of two-ring purines (A <> G) or of one-ring pyrimidines (C <> T)</li>
<li><em>Transversions</em> are interchanges of purine for pyrimidine bases or vice versa (A <> T | A <> C | G <> T | G <> C)</li>
</ul>
</li>
</ul>
<img style="max-width:30%" src="../img/transitions-transversions.svg"></img>
<br>
<hr />
<h4><a name="Proportional-distance" class="anchor" href="#Proportional-distance">Proportional distance</a></h4>
<p>The pairwise proportional distance (or <em>p distance</em>) is the classical 'naive' approach to estimate pairwise distances between two sequences.</p>
<p>It is simply the ratio between substitution sites and the length of the sequences. Note that - as in all distances you will implement - both
sequences have to be of the same length or have to be aligned before calculating distances. It is obtained by dividing the amount of substitutions
by the total amount of compared nucleotides:</p>
<p><span class="math">\(p = \frac{d}{L},\)</span></p>
<p>where</p>
<p><span class="math">\(d = s + v\)</span></p>
<p><span class="math">\(s : Transitions\)</span></p>
<p><span class="math">\(v : Transversions\)</span></p>
<p><span class="math">\(L : length\)</span></p>
<p>What are the advantages and weaknesses of this approach? Inform yourself.</p>
<br>
<hr />
<h3><a name="Distance-Corrections-based-on-evolutionary-Models" class="anchor" href="#Distance-Corrections-based-on-evolutionary-Models">Distance Corrections based on evolutionary Models</a></h3>
<p>To overcome the shortcomings of the simple proportional distance, many evolutionary models for biological sequences are available.
Most of the time, sequence evolution by mutation is described as a stochastic process modelled by continuous-time Markov chains, with
the alphabet (for DNA: <em>A, C, G, T</em>) as possible states.</p>
<p>For any sequence position, these models are then defining a substitution probability matrix based that acts as the stochastic matrix of the markov chain.</p>
<p>A basic understanding of markov chains may be beneficial, but not necessary. It is just important that these models define probabilities
for transitions and transversions, and can be solved for a corrected pairwise distance that suffices the model substitution criteria.</p>
<br>
<hr />
<h4><a name="The-Jukes-Cantor-Model" class="anchor" href="#The-Jukes-Cantor-Model">The Jukes-Cantor Model</a></h4>
<p>The Jukes-Cantor model is the simplest form of these kinds of models. It makes no difference between transitions and transversions,
meaning that all substitutions have the same substitution rate (<span class="math">\(\alpha\)</span>). Also, all sites are modelled independently.</p>
<p>The substitution matrix is:</p>
<p><span class="math">\[\begin{matrix}
&amp; \begin{matrix}A &amp; &amp; C &amp; &amp; G &amp; &amp; T\end{matrix} \\\\
\begin{matrix}A\\\\C\\\\G\\\\T\end{matrix} &amp; 
  \begin{pmatrix}
    -3\alpha&amp;\alpha&amp;\alpha&amp;\alpha\\\\
    \alpha&amp;-3\alpha&amp;\alpha&amp;\alpha\\\\
    \alpha&amp;\alpha&amp;-3\alpha&amp;\alpha\\\\
    \alpha&amp;\alpha&amp;\alpha&amp;-3\alpha
  \end{pmatrix}\\\\
\end{matrix}\]</span></p>
<p><span class="math">\(3\alpha t\)</span> mutations would be expected during a time <span class="math">\(t\)</span> for each sequence site on each sequence, leading to
a correction factor for the proportional distance <span class="math">\(d_{JC}\)</span> :</p>
<p><span class="math">\[d_{JC}=-\frac{3}{4}ln(1-\frac{4}{3}p)\]</span></p>
<p>What are the advantages and weaknesses of this approach? Inform yourself.</p>
<br>
<hr />
<h4><a name="The-Kimura-2-Parameter-Model" class="anchor" href="#The-Kimura-2-Parameter-Model">The Kimura 2-Parameter Model</a></h4>
<p>The substitution matrix is:</p>
<p><span class="math">\[\begin{matrix}
&amp; \begin{matrix}A &amp; &amp; &amp; &amp; C &amp; &amp; &amp; &amp; G &amp; &amp; &amp; &amp; T\end{matrix} \\\\
\begin{matrix}A\\\\C\\\\G\\\\T\end{matrix} &amp; 
  \begin{pmatrix}
    -2\beta-\alpha&amp;\beta&amp;\alpha&amp;\beta\\\\
    \beta&amp;-2\beta-\alpha&amp;\beta&amp;\alpha\\\\
    \alpha&amp;\beta&amp;-2\beta-\alpha&amp;\beta\\\\
    \beta&amp;\alpha&amp;\beta&amp;-2\beta-\alpha
  \end{pmatrix}\\\\
\end{matrix}\]</span></p>
<p>It results in a corrected distance <span class="math">\(d_{K2P}\)</span>:</p>
<p><span class="math">\(d_{K2P}=-\frac{1}{2}ln(1-2P-Q)-\frac{1}{4}ln(1-2Q)\)</span>,</p>
<p>where</p>
<p><span class="math">\(P=\frac{s}{L}\)</span></p>
<p><span class="math">\(Q=\frac{v}{L}\)</span></p>
<p>What are the advantages and weaknesses of this approach? Inform yourself.</p>
<br>
<hr />
<h2><a name="Aim-for-this-project" class="anchor" href="#Aim-for-this-project">Aim for this project</a></h2>
<ul>
<li>
<p>You understand the following evolutionary distance models and are able to explain the differences between them (required in your final report)
Also, you implement them in F# for the BioFSharp library:</p>
<ul>
<li>Pairwise p Distance</li>
<li>JC69 evolutionary distance based on the model by Jukes and Cantor</li>
<li>K81 evolutionary distance based on the Kimura two-parameter model</li>
</ul>
</li>
<li>
<p>As a demonstration of your implementations, as well as to show the differences between these models,
You construct at least 10 adequate test sequences of equal length, and construct phyologenetic trees from them.
Investigate the most interesting and obvious differences, and relate them to the different model assumptions.</p>
</li>
<li>Finally, you choose adequate sequences of at least 6 organisms, perform a multiple alignment for them and repeat above process for real-world sequences.</li>
<li>You present above results in a blog post resembling your final report.</li>
<li><p><strong>Bonus</strong>: You implement a visualization method for phylogenetic trees for <a href="https://fslab.org/Cyjs.NET/">Cyjs.NET</a>.</p></li>
</ul>
<h2><a name="Coding-clues" class="anchor" href="#Coding-clues">Coding clues</a></h2>
<h3><a name="Before-you-start" class="anchor" href="#Before-you-start">Before you start</a></h3>
<ul>
<li>(re)familiarize yourself with the basics behind phylogenetic trees.</li>
<li>(re)familiarize yourself with F# function signatures and the basics of F# programming.</li>
<li>A basic understanding of Markov chains is beneficial, but not necessary.</li>
</ul>
<h3><a name="Scripting-environment-and-necessary-libraries" class="anchor" href="#Scripting-environment-and-necessary-libraries">Scripting environment and necessary libraries</a></h3>
<ul>
<li>
create a F# script (.fsx), load and open the following libraries:
<ul>
<li><code>FSharpAux</code></li>
<li><code>FSharpAux.IO</code></li>
<li><code>FSharp.Stats</code></li>
<li><code>BioFSharp</code></li>
<li><code>BioFSharp.IO</code></li>
<li><code>Plotly.NET</code></li>
</ul>
</li>
</ul>
<p>the top of your script file should look like this:</p>
<pre class="fssnip highlighted"><code lang="fsharp"><span class="pp">#r</span> <span class="s">&quot;nuget: FSharpAux&quot;</span>
<span class="pp">#r</span> <span class="s">&quot;nuget: FSharpAux.IO&quot;</span>
<span class="pp">#r</span> <span class="s">&quot;nuget: FSharp.Stats, 0.4.1&quot;</span> 
<span class="pp">#r</span> <span class="s">&quot;nuget: BioFSharp, 2.0.0-beta6&quot;</span>
<span class="pp">#r</span> <span class="s">&quot;nuget: BioFSharp.IO, 2.0.0-beta6&quot;</span>
<span class="pp">#r</span> <span class="s">&quot;nuget: Plotly.NET, 2.0.0-beta9&quot;</span>

<span class="k">open</span> <span onmouseout="hideTip(event, 'fs1', 1)" onmouseover="showTip(event, 'fs1', 1)" class="id">FSharpAux</span>
<span class="k">open</span> <span onmouseout="hideTip(event, 'fs1', 2)" onmouseover="showTip(event, 'fs1', 2)" class="id">FSharpAux</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs2', 3)" onmouseover="showTip(event, 'fs2', 3)" class="id">IO</span>
<span class="k">open</span> <span onmouseout="hideTip(event, 'fs3', 4)" onmouseover="showTip(event, 'fs3', 4)" class="id">BioFSharp</span>
<span class="k">open</span> <span onmouseout="hideTip(event, 'fs3', 5)" onmouseover="showTip(event, 'fs3', 5)" class="id">BioFSharp</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs4', 6)" onmouseover="showTip(event, 'fs4', 6)" class="id">IO</span>
<span class="k">open</span> <span onmouseout="hideTip(event, 'fs5', 7)" onmouseover="showTip(event, 'fs5', 7)" class="id">FSharp</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs6', 8)" onmouseover="showTip(event, 'fs6', 8)" class="id">Stats</span>
<span class="k">open</span> <span onmouseout="hideTip(event, 'fs7', 9)" onmouseover="showTip(event, 'fs7', 9)" class="id">Plotly</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs8', 10)" onmouseover="showTip(event, 'fs8', 10)" class="id">NET</span>
</code></pre>
<h3><a name="General-coding-advice" class="anchor" href="#General-coding-advice">General coding advice</a></h3>
<ul>
<li>
All pairwise distance functions should be able to operate on either <code>BioArray</code>, <code>BioList</code>, or <code>BioSeq</code>. 
You can use the fact that all of these sequence types are implementing <code>IEnumerable</code> and can only contain nucleotides.
</li>
</ul>
<br>
<hr />
<ul>
<li>
To perform hierarchical clustering to reconstruct the phylogenetic trees, use the respective module from the <code>BioFSharp</code> library 
(<code>PhylogeneticTree</code> functions) and work with <code>TaggedSequence</code>s:
</li>
</ul>
<br>
<pre class="fssnip highlighted"><code lang="fsharp"><span class="k">open</span> <span onmouseout="hideTip(event, 'fs3', 11)" onmouseover="showTip(event, 'fs3', 11)" class="id">BioFSharp</span>
</code></pre>
<pre class="fssnip highlighted"><code lang="fsharp"><span class="k">let</span> <span onmouseout="hideTip(event, 'fs9', 12)" onmouseover="showTip(event, 'fs9', 12)" class="fn">yourDistance</span> <span class="pn">(</span><span onmouseout="hideTip(event, 'fs10', 13)" onmouseover="showTip(event, 'fs10', 13)" class="fn">seqA</span><span class="pn">:</span> <span onmouseout="hideTip(event, 'fs11', 14)" onmouseover="showTip(event, 'fs11', 14)" class="if">seq</span><span class="pn">&lt;</span><span onmouseout="hideTip(event, 'fs12', 15)" onmouseover="showTip(event, 'fs12', 15)" class="if">IBioItem</span><span class="pn">&gt;</span><span class="pn">)</span> <span class="pn">(</span><span onmouseout="hideTip(event, 'fs13', 16)" onmouseover="showTip(event, 'fs13', 16)" class="fn">seqB</span><span class="pn">:</span><span onmouseout="hideTip(event, 'fs11', 17)" onmouseover="showTip(event, 'fs11', 17)" class="if">seq</span><span class="pn">&lt;</span><span onmouseout="hideTip(event, 'fs12', 18)" onmouseover="showTip(event, 'fs12', 18)" class="if">IBioItem</span><span class="pn">&gt;</span><span class="pn">)</span> <span class="o">=</span> <span class="o">..</span><span class="pn">.</span>
</code></pre>
<pre class="fssnip highlighted"><code lang="fsharp"><span class="c">//reconstruct a phylogenetic tree from tagged sequences</span>
<span class="k">let</span> <span onmouseout="hideTip(event, 'fs14', 19)" onmouseover="showTip(event, 'fs14', 19)" class="id">myTree</span> <span class="o">=</span>
    <span onmouseout="hideTip(event, 'fs15', 20)" onmouseover="showTip(event, 'fs15', 20)" class="rt">PhylogeneticTree</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs16', 21)" onmouseover="showTip(event, 'fs16', 21)" class="id">ofTaggedBioSequences</span>
        <span onmouseout="hideTip(event, 'fs9', 22)" onmouseover="showTip(event, 'fs9', 22)" class="fn">yourDistance</span> <span class="c">// your distance function for either p, JC69, or K81 distance</span>
        <span class="id">yourSequences</span> <span class="c">// your adequate nucleotide test sequences as tagged sequences </span>
</code></pre>
<br>
<hr />
<ul>
<li>
to perform Multiple sequence alignment between your real world sequence examples, use the respective functions from <code>BioFSharp</code>:

  For that, you have to first install <a href="http://www.clustal.org/omega/clustal-omega-1.2.2-win64.zip">clustal omega</a>, a very nice multiple sequence aligment tool. 
  You can use it directly in F# interactive via BioFSharp's <code>ClustalOWrapper</code>:
</li>
</ul>
<br>
<pre class="fssnip highlighted"><code lang="fsharp"><span class="k">open</span> <span onmouseout="hideTip(event, 'fs17', 23)" onmouseover="showTip(event, 'fs17', 23)" class="m">ClustalOWrapper</span>
<span class="k">let</span> <span onmouseout="hideTip(event, 'fs18', 24)" onmouseover="showTip(event, 'fs18', 24)" class="id">cw</span> <span class="o">=</span> <span onmouseout="hideTip(event, 'fs19', 25)" onmouseover="showTip(event, 'fs19', 25)" class="fn">ClustalOWrapper</span><span class="pn">(</span><span class="s">&quot;path/where/you/extracted/clustal-omega-1.2.2-win64/clustalo.exe&quot;</span><span class="pn">)</span> <span class="c">// replace with real path from your machine here!</span>

<span class="k">let</span> <span onmouseout="hideTip(event, 'fs20', 26)" onmouseover="showTip(event, 'fs20', 26)" class="id">sequences</span> <span class="o">=</span> 
    <span class="pn">[</span>
    <span onmouseout="hideTip(event, 'fs21', 27)" onmouseover="showTip(event, 'fs21', 27)" class="rt">TaggedSequence</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs22', 28)" onmouseover="showTip(event, 'fs22', 28)" class="id">create</span> <span class="s">&quot;seq1&quot;</span> <span class="pn">(</span><span class="s">&quot;ATGAAAAA&quot;</span><span class="pn">)</span>
    <span onmouseout="hideTip(event, 'fs21', 29)" onmouseover="showTip(event, 'fs21', 29)" class="rt">TaggedSequence</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs22', 30)" onmouseover="showTip(event, 'fs22', 30)" class="id">create</span> <span class="s">&quot;seq2&quot;</span> <span class="pn">(</span><span class="s">&quot;ATGAAACA&quot;</span><span class="pn">)</span>
    <span onmouseout="hideTip(event, 'fs21', 31)" onmouseover="showTip(event, 'fs21', 31)" class="rt">TaggedSequence</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs22', 32)" onmouseover="showTip(event, 'fs22', 32)" class="id">create</span> <span class="s">&quot;seq3&quot;</span> <span class="pn">(</span><span class="s">&quot;ATGAAAAAAT&quot;</span><span class="pn">)</span>
    <span onmouseout="hideTip(event, 'fs21', 33)" onmouseover="showTip(event, 'fs21', 33)" class="rt">TaggedSequence</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs22', 34)" onmouseover="showTip(event, 'fs22', 34)" class="id">create</span> <span class="s">&quot;seq4&quot;</span> <span class="pn">(</span><span class="s">&quot;ATGGAAAA&quot;</span><span class="pn">)</span>
    <span class="pn">]</span>

<span class="k">let</span> <span onmouseout="hideTip(event, 'fs23', 35)" onmouseover="showTip(event, 'fs23', 35)" class="id">alignedSequences</span> <span class="o">=</span> 
    <span onmouseout="hideTip(event, 'fs18', 36)" onmouseover="showTip(event, 'fs18', 36)" class="id">cw</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs24', 37)" onmouseover="showTip(event, 'fs24', 37)" class="id">AlignSequences</span><span class="pn">(</span><span onmouseout="hideTip(event, 'fs20', 38)" onmouseover="showTip(event, 'fs20', 38)" class="id">sequences</span><span class="pn">,</span><span onmouseout="hideTip(event, 'fs25', 39)" onmouseover="showTip(event, 'fs25', 39)" class="m">Seq</span><span class="pn">.</span><span onmouseout="hideTip(event, 'fs26', 40)" onmouseover="showTip(event, 'fs26', 40)" class="id">empty</span><span class="pn">)</span>
</code></pre>
<br>
<hr />
<ul>
<li>
There are two ways of handling the gaps produced by alignments: <em>Complete-Deletion</em> and <em>Pairwise Deletion</em> inform yourself about them.<br />

  <em>Hint1</em>: The output of above alignment contains conservation information about the alignment. 

  <em>Hint2</em>: You might want to add an additional parameter to your distance function modelling this behaviour. For that, use a Discriminated Union type.
</li>
</ul>
<br>
<hr />
<ul>
<li>Here is a suggestion for the general workflow for the real-world sequences:</li>
</ul>
<blockquote>
<p><strong>Suggested workflow</strong></p>
<p>Read your fasta formatted sequences</p>
<p>Perform multiple sequence alignment <span class="math">\(A\)</span></p>
<p>for each implemented distance function <span class="math">\(dist\)</span>:</p>
<p><span class="math">\(\quad\)</span>reconstruct phylogenetic tree for <span class="math">\(A\)</span> with <span class="math">\(dist\)</span> function</p>
<p><span class="math">\(\quad\)</span>Write as newick format</p>
<p><span class="math">\(\quad\)</span>Visualize tree, for example on <a href="http://etetoolkit.org/treeview/">http://etetoolkit.org/treeview/</a></p>
</blockquote>
<hr />
<hr />
<h2><a name="References" class="anchor" href="#References">References</a></h2>
<p>Nei, M. &amp; Zhang J. Evolutionary Distance: Estimation 2006 <a href="https://doi.org/10.1038/npg.els.0005108">https://doi.org/10.1038/npg.els.0005108</a></p>
<p><a href="https://en.wikipedia.org/wiki/Models_of_DNA_evolution">https://en.wikipedia.org/wiki/Models_of_DNA_evolution</a></p>
<p><a href="https://www.cs.rice.edu/~nakhleh/COMP571/Slides/Phylogenetics-DistanceMethods-Full.pdf">https://www.cs.rice.edu/~nakhleh/COMP571/Slides/Phylogenetics-DistanceMethods-Full.pdf</a></p>
<p><a href="https://www.megasoftware.net/web_help_10/index.htm#t=Models_for_estimating_distances.htm">https://www.megasoftware.net/web_help_10/index.htm#t=Models_for_estimating_distances.htm</a></p>
<p><a href="https://www.megasoftware.net/mega1_manual/Distance.html">https://www.megasoftware.net/mega1_manual/Distance.html</a></p>
<h2><a name="Additional-information" class="anchor" href="#Additional-information">Additional information</a></h2>
<h3><a name="Testing" class="anchor" href="#Testing">Testing</a></h3>
<ul>
<li>
<p>the <a href="https://www.megasoftware.net/active_download">MEGA software suit</a> contains many evolutionary distance functions.
You can use it as reference implementation to check wether your functions return correct results. To do that, download the GUI version, install it,
and follow these steps:</p>
<ul>
<li>
<p>Click the <code>Distances</code> button and choose <code>pairwise distance</code>:</p>
<p><img src="../img/mega1.png" alt="" /></p>
</li>
<li>
<p>Select a fasta file containing the sequences of interest, and choose the reference model from this dialogue:</p>
<p><img src="../img/mega2.png" alt="" /></p>
</li>
</ul>
</li>
</ul>
<h3><a name="Blog-post" class="anchor" href="#Blog-post">Blog post</a></h3>
<ul>
<li>Donâ€™t forget to describe the limits/weaknesses of the models in your blog post.</li>
<li>How to handle gaps in aligned sequences?</li>
<li>Provide reasoning why you chose your real world sequences</li>
</ul>

                </div>
            </div>
        </div>
        <div class="fsdocs-tip" id="fs1">namespace FSharpAux</div>
<div class="fsdocs-tip" id="fs2">namespace FSharpAux.IO</div>
<div class="fsdocs-tip" id="fs3">namespace BioFSharp</div>
<div class="fsdocs-tip" id="fs4">namespace BioFSharp.IO</div>
<div class="fsdocs-tip" id="fs5">Multiple items<br />namespace FSharp<br /><br />--------------------<br />namespace Microsoft.FSharp</div>
<div class="fsdocs-tip" id="fs6">namespace FSharp.Stats</div>
<div class="fsdocs-tip" id="fs7">namespace Plotly</div>
<div class="fsdocs-tip" id="fs8">namespace Plotly.NET</div>
<div class="fsdocs-tip" id="fs9">val yourDistance : seqA:seq&lt;IBioItem&gt; -&gt; seqB:seq&lt;IBioItem&gt; -&gt; &#39;a</div>
<div class="fsdocs-tip" id="fs10">val seqA : seq&lt;IBioItem&gt;</div>
<div class="fsdocs-tip" id="fs11">Multiple items<br />val seq : sequence:seq&lt;&#39;T&gt; -&gt; seq&lt;&#39;T&gt;<br /><em>&lt;summary&gt;Builds a sequence using sequence expression syntax&lt;/summary&gt;<br />&lt;param name=&quot;sequence&quot;&gt;The input sequence.&lt;/param&gt;<br />&lt;returns&gt;The result sequence.&lt;/returns&gt;</em><br /><br />--------------------<br />type seq&lt;&#39;T&gt; = System.Collections.Generic.IEnumerable&lt;&#39;T&gt;<br /><em>&lt;summary&gt;An abbreviation for the CLI type &lt;see cref=&quot;T:System.Collections.Generic.IEnumerable`1&quot; /&gt;&lt;/summary&gt;<br />&lt;remarks&gt;
  See the &lt;see cref=&quot;T:Microsoft.FSharp.Collections.SeqModule&quot; /&gt; module for further operations related to sequences.

  See also &lt;a href=&quot;https://docs.microsoft.com/dotnet/fsharp/language-reference/sequences&quot;&gt;F# Language Guide - Sequences&lt;/a&gt;.
&lt;/remarks&gt;</em></div>
<div class="fsdocs-tip" id="fs12">type IBioItem =
  abstract member Formula : Formula
  abstract member Name : string
  abstract member Symbol : char
  abstract member isGap : bool
  abstract member isTerminator : bool<br /><em>&lt;summary&gt;
Marker interface for BioItem base.
&lt;/summary&gt;</em></div>
<div class="fsdocs-tip" id="fs13">val seqB : seq&lt;IBioItem&gt;</div>
<div class="fsdocs-tip" id="fs14">val myTree : PhylogeneticTree&lt;TaggedSequence&lt;string,IBioItem&gt; * string&gt;</div>
<div class="fsdocs-tip" id="fs15">type PhylogeneticTree&lt;&#39;T&gt; =
  | Branch of &#39;T * List&lt;PhylogeneticTree&lt;&#39;T&gt;&gt;
    static member addChildToNodes : condition:(PhylogeneticTree&lt;&#39;T&gt; -&gt; bool) -&gt; child:PhylogeneticTree&lt;&#39;T&gt; -&gt; tree:PhylogeneticTree&lt;&#39;T&gt; -&gt; PhylogeneticTree&lt;&#39;T&gt;
    static member countLeafs : tree:PhylogeneticTree&lt;&#39;T&gt; -&gt; int
    static member fold : acc:&#39;State -&gt; folder:(&#39;State -&gt; PhylogeneticTree&lt;&#39;T&gt; -&gt; &#39;State) -&gt; tree:PhylogeneticTree&lt;&#39;T&gt; -&gt; &#39;State
    static member iter : action:(PhylogeneticTree&lt;&#39;T&gt; -&gt; unit) -&gt; tree:PhylogeneticTree&lt;&#39;T&gt; -&gt; unit
    static member map : mapping:(PhylogeneticTree&lt;&#39;T&gt; -&gt; &#39;t) -&gt; tree:PhylogeneticTree&lt;&#39;T&gt; -&gt; PhylogeneticTree&lt;&#39;t&gt;
    static member mapFold : acc:&#39;State -&gt; mapping:(PhylogeneticTree&lt;&#39;T&gt; -&gt; &#39;t) -&gt; folder:(&#39;State -&gt; &#39;t -&gt; &#39;State) -&gt; tree:PhylogeneticTree&lt;&#39;T&gt; -&gt; &#39;State
    static member ofHierarchicalCluster : branchTag:&#39;T -&gt; distanceConverter:(float -&gt; &#39;Distance) -&gt; hCluster:Cluster&lt;&#39;T&gt; -&gt; PhylogeneticTree&lt;&#39;T * &#39;Distance&gt;
    static member ofTaggedBioSequences : distanceFunction:(seq&lt;&#39;a1&gt; -&gt; seq&lt;&#39;a1&gt; -&gt; float) -&gt; sequences:seq&lt;TaggedSequence&lt;string,&#39;a1&gt;&gt; -&gt; PhylogeneticTree&lt;TaggedSequence&lt;string,&#39;a1&gt; * string&gt; (requires &#39;a1 :&gt; IBioItem)
    static member ofTaggedSequencesWithLinker : branchTag:&#39;T -&gt; distanceConverter:(float -&gt; &#39;Distance) -&gt; distanceFunction:(seq&lt;&#39;S&gt; -&gt; seq&lt;&#39;S&gt; -&gt; float) -&gt; linker:LancWilliamsLinker -&gt; sequences:seq&lt;TaggedSequence&lt;&#39;T,&#39;S&gt;&gt; -&gt; PhylogeneticTree&lt;TaggedSequence&lt;&#39;T,&#39;S&gt; * &#39;Distance&gt;
    static member tryGetNodeBy : condition:(PhylogeneticTree&lt;&#39;T&gt; -&gt; bool) -&gt; tree:PhylogeneticTree&lt;&#39;T&gt; -&gt; PhylogeneticTree&lt;&#39;T&gt; option<br /><em>&lt;summary&gt;
 Recursive representation of a phylogenetic tree
&lt;/summary&gt;</em></div>
<div class="fsdocs-tip" id="fs16">static member PhylogeneticTree.ofTaggedBioSequences : distanceFunction:(seq&lt;&#39;a1&gt; -&gt; seq&lt;&#39;a1&gt; -&gt; float) -&gt; sequences:seq&lt;TaggedSequence&lt;string,&#39;a1&gt;&gt; -&gt; PhylogeneticTree&lt;TaggedSequence&lt;string,&#39;a1&gt; * string&gt; (requires &#39;a1 :&gt; IBioItem)</div>
<div class="fsdocs-tip" id="fs17">module ClustalOWrapper

from BioFSharp.IO<br /><em>&lt;summary&gt;
Wrapper and its helpers for Clustal Omega multiple alignment tools
&lt;/summary&gt;</em></div>
<div class="fsdocs-tip" id="fs18">val cw : ClustalOWrapper</div>
<div class="fsdocs-tip" id="fs19">Multiple items<br />module ClustalOWrapper

from BioFSharp.IO<br /><em>&lt;summary&gt;
Wrapper and its helpers for Clustal Omega multiple alignment tools
&lt;/summary&gt;</em><br /><br />--------------------<br />type ClustalOWrapper =
  new : ?rootPath:string -&gt; ClustalOWrapper
  member AlignFromFile : inputPath:Input * outputPath:string * parameters:seq&lt;ClustalParams&gt; * ?name:string -&gt; unit
  member AlignSequences : input:seq&lt;TaggedSequence&lt;string,char&gt;&gt; * parameters:seq&lt;ClustalParams&gt; * ?name:string -&gt; Alignment&lt;TaggedSequence&lt;string,char&gt;,AlignmentInfo&gt;<br /><em>&lt;summary&gt;
 A wrapper to perform Clustal Omega alignment tasks    
&lt;/summary&gt;</em><br /><br />--------------------<br />new : ?rootPath:string -&gt; ClustalOWrapper</div>
<div class="fsdocs-tip" id="fs20">val sequences : TaggedSequence&lt;string,char&gt; list</div>
<div class="fsdocs-tip" id="fs21">type TaggedSequence&lt;&#39;T,&#39;S&gt; =
  { Tag: &#39;T
    Sequence: seq&lt;&#39;S&gt; }
    static member create : tag:&#39;T -&gt; sequence:seq&lt;&#39;S&gt; -&gt; TaggedSequence&lt;&#39;T,&#39;S&gt;
    static member mapSequence : mapping:(seq&lt;&#39;S&gt; -&gt; seq&lt;&#39;M&gt;) -&gt; ts:TaggedSequence&lt;&#39;T,&#39;S&gt; -&gt; TaggedSequence&lt;&#39;T,&#39;M&gt;
    static member mapTag : mapping:(&#39;T -&gt; &#39;U) -&gt; ts:TaggedSequence&lt;&#39;T,&#39;S&gt; -&gt; TaggedSequence&lt;&#39;U,&#39;S&gt;<br /><em>&lt;summary&gt;
 Record of a sequence and its tag
&lt;/summary&gt;</em></div>
<div class="fsdocs-tip" id="fs22">static member TaggedSequence.create : tag:&#39;T -&gt; sequence:seq&lt;&#39;S&gt; -&gt; TaggedSequence&lt;&#39;T,&#39;S&gt;</div>
<div class="fsdocs-tip" id="fs23">val alignedSequences : Alignment.Alignment&lt;TaggedSequence&lt;string,char&gt;,Clustal.AlignmentInfo&gt;</div>
<div class="fsdocs-tip" id="fs24">member ClustalOWrapper.AlignSequences : input:seq&lt;TaggedSequence&lt;string,char&gt;&gt; * parameters:seq&lt;Parameters.ClustalParams&gt; * ?name:string -&gt; Alignment.Alignment&lt;TaggedSequence&lt;string,char&gt;,Clustal.AlignmentInfo&gt;</div>
<div class="fsdocs-tip" id="fs25">Multiple items<br />module Seq

from Plotly.NET<br /><br />--------------------<br />module Seq

from FSharp.Stats<br /><em>&lt;summary&gt;
 Module to compute common statistical measure
&lt;/summary&gt;</em><br /><br />--------------------<br />module Seq

from FSharpAux<br /><br />--------------------<br />module Seq

from Microsoft.FSharp.Collections<br /><em>&lt;summary&gt;Contains operations for working with values of type &lt;see cref=&quot;T:Microsoft.FSharp.Collections.seq`1&quot; /&gt;.&lt;/summary&gt;</em><br /><br />--------------------<br />type Seq =
  static member CSV : separator:string -&gt; header:bool -&gt; flatten:bool -&gt; data:seq&lt;&#39;a&gt; -&gt; seq&lt;string&gt;
  static member CSVwith : valFunc:(&#39;a -&gt; (&#39;a -&gt; obj) []) -&gt; strFunc:(string -&gt; bool -&gt; obj -&gt; obj -&gt; string) -&gt; separator:string -&gt; header:bool -&gt; flatten:bool -&gt; data:seq&lt;&#39;a&gt; -&gt; seq&lt;string&gt;
  static member fromFile : filePath:string -&gt; seq&lt;string&gt;
  static member fromFileWithCsvSchema : filePath:string * separator:char * firstLineHasHeader:bool * ?skipLines:int * ?skipLinesBeforeHeader:int * ?schemaMode:SchemaModes -&gt; seq&lt;&#39;schema&gt;
  static member fromFileWithSep : separator:char -&gt; filePath:string -&gt; seq&lt;string []&gt;
  static member stringFunction : separator:string -&gt; flatten:bool -&gt; input:&#39;a -&gt; (obj -&gt; string)
  static member valueFunction : dataEntry:&#39;a -&gt; (&#39;a -&gt; obj) []
  static member write : path:string -&gt; data:seq&lt;&#39;a&gt; -&gt; unit
  static member writeOrAppend : path:string -&gt; data:seq&lt;&#39;a&gt; -&gt; unit</div>
<div class="fsdocs-tip" id="fs26">val empty&lt;&#39;T&gt; : seq&lt;&#39;T&gt;<br /><em>&lt;summary&gt;Creates an empty sequence.&lt;/summary&gt;<br />&lt;returns&gt;An empty sequence.&lt;/returns&gt;</em></div>

    </div>

    <!-- BEGIN SEARCH BOX: this adds support for the search box -->
    <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/JavaScript-autoComplete/1.0.4/auto-complete.css" />
    <script type="text/javascript">var fsdocs_search_baseurl = '/BIO-BTE-10-L-4/'</script>
    <script type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/lunr.js/2.3.8/lunr.min.js"></script>
    <script type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/JavaScript-autoComplete/1.0.4/auto-complete.min.js"></script>
    <script type="text/javascript" src="/BIO-BTE-10-L-4/content/fsdocs-search.js"></script>
    <!-- END SEARCH BOX: this adds support for the search box -->
</body>
<footer>
    <script>
        bulmaCollapsible.attach('.is-collapsible');
    </script>
</footer>
</html>