<!DOCTYPE html>
<html lang="en">

<head>
    <meta charset="utf-8">
    <title>Molecular evolution of thermolabile proteins in microalgae</title>
    <meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
    <meta name="author" content="">
    <link rel="shortcut icon" type="image/x-icon" href="/BIO-BTE-10-L-4//img/favicon.ico">
    <script src="https://code.jquery.com/jquery-3.4.1.min.js"></script>
    <script type="text/javascript" src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <!-- <script type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.7/MathJax.js?config=TeX-MML-AM_CHTML"></script> -->

    <link type="text/css" rel="stylesheet" href="/BIO-BTE-10-L-4/content/fsdocs-default.css" />
    <link type="text/css" rel="stylesheet" href="/BIO-BTE-10-L-4/content/fsdocs-custom.css" />
    <script type="text/javascript" src="/BIO-BTE-10-L-4/content/fsdocs-tips.js"></script>
    <script type="text/javascript" src="/BIO-BTE-10-L-4/content/init.js"></script>
    <script type="text/javascript" src="https://cdn.jsdelivr.net/npm/@creativebulma/bulma-collapsible@1.0.4/dist/js/bulma-collapsible.min.js"></script>
    <link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/@creativebulma/bulma-collapsible@1.0.4/dist/css/bulma-collapsible.min.css">
    <!-- HTML5 shim, for IE6-8 support of HTML5 elements -->
    <!--[if lt IE 9]>
      <script src="http://html5shim.googlecode.com/svn/trunk/html5.js"></script>
    <![endif]-->
    <!-- BEGIN SEARCH BOX: this adds support for the search box -->
    <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/JavaScript-autoComplete/1.0.4/auto-complete.css" />
    <!-- END SEARCH BOX: this adds support for the search box -->
    
</head> 

<body>
    <div class="columns is-fullheight m-0">
        <div id="fsdocs-menu" class="column is-2 is-paddingless">
            <aside class="menu">
                <div class="container is-paddingless is-marginless has-background-dark">
                    <a href="/BIO-BTE-10-L-4/">
                        <figure class="image is-128x128 container">
                            <img src="/BIO-BTE-10-L-4/img/logo.png">
                        </figure>
                    </a>
                    <h2 class="title has-text-centered has-text-white pb-4">BIO-BTE-10-L-4</h2>
                </div>
                <div class="content">
                    <!-- BEGIN SEARCH BOX: this adds support for the search box -->
                    <div id="header">
                        <div class="field">
                            <div class="control has-icons-left has-icons-right searchbox" id="fsdocs-searchbox">
                                <input data-search-input="" id="search-by" type="search" placeholder="Search..." class="input"/>
                                <span class="icon is-small is-left">
                                    <i class="fas fa-search"></i>
                                </span>
                            </div>
                        </div>
                    </div>
                    <!-- END SEARCH BOX: this adds support for the search box -->
                    <ul class="navbar-nav">
                        <li class="nav-header">
  general
</li>             
<li class="nav-item">
  <a class="nav-link" href="/BIO-BTE-10-L-4/general/example-blog-post.html">
    Example blog post
  </a>
</li>             
<li class="nav-item">
  <a class="nav-link" href="/BIO-BTE-10-L-4/general/blog-post-guide.html">
    Blog post guide
  </a>
</li>             
<li class="nav-header">
  projects
</li>             
<li class="nav-item">
  <a class="nav-link" href="/BIO-BTE-10-L-4/projects/tSNE.html">
    t-Distributed Stochastic Neighbour Embedding
  </a>
</li>             
<li class="nav-item">
  <a class="nav-link" href="/BIO-BTE-10-L-4/projects/hClust.html">
    Efficient agglomerative hierarchical clustering
  </a>
</li>             
<li class="nav-item">
  <a class="nav-link" href="/BIO-BTE-10-L-4/projects/spectral-alignment.html">
    Alignment of Peptide derived MS Spectra
  </a>
</li>             
<li class="nav-item">
  <a class="nav-link" href="/BIO-BTE-10-L-4/projects/evolutionary-distance.html">
    Phylogenetic tree reconstruction based on evolutionary distance
  </a>
</li>             
<li class="nav-item">
  <a class="nav-link" href="/BIO-BTE-10-L-4/projects/PrimsAlgorithm.html">
    Reduce complex graphs to the best paths
  </a>
</li>             
<li class="nav-item">
  <a class="nav-link" href="/BIO-BTE-10-L-4/projects/sequencingByHybridization.html">
    Sequencing by Hybridization as an Eulerian Path Problem
  </a>
</li>             
<li class="nav-header">
  bachelor thesis
</li>             
<li class="nav-item">
  <a class="nav-link" href="/BIO-BTE-10-L-4/bachelor/molecular-evolution.html">
    Molecular evolution of thermolabile proteins in microalgae
  </a>
</li>
                    </ul>
                </div>
            </aside>
        </div>
        <div class="column is-10 is-paddingless">
            <div class="section pt-4">
                <div class="container" id="fsdocs-content">
                    
<h1><a name="Molecular-evolution-of-thermolabile-proteins-in-microalgae" class="anchor" href="#Molecular-evolution-of-thermolabile-proteins-in-microalgae">Molecular evolution of thermolabile proteins in microalgae</a></h1>
<p><strong>Interested?</strong> Contact <a href="mailto:muehlhaus@bio.uni-kl.de">muehlhaus@bio.uni-kl.de</a> or <a href="mailto:venn@bio.uni-kl.de">venn@bio.uni-kl.de</a></p>
<h3><a name="Table-of-contents" class="anchor" href="#Table-of-contents">Table of contents</a></h3>
<ul>
<li><a href="#Introduction">Introduction</a></li>
<li><a href="#Description">Description</a></li>
<li><a href="#Additional-information">Additional information</a></li>
<li><a href="#References">References</a></li>
</ul>
<h1><a name="Introduction" class="anchor" href="#Introduction">Introduction</a></h1>
<p>The bachelor thesis focuses on the analysis of evolutionary constrained proteins. It aims to investigate the evolutionary rates of proteins and identify a special kind of key-regulator-mechanism in acclimation responses.</p>
<p>If you think of evolution, you intentionally assume, that sequences tend to evolve to become more thermostable over time.
But what is the reason that some sequences seems to be hindered to evolve and are kept thermolabile?
The idea is that defolding or aggregation during heat stress is no damage at all events, but more like a mechanism to trigger some heat stress responses.
Of course, this is not true for all aggregating proteins but maybe for some master key-regulators which have switch functions to (de)activate specific pathways if the cellular pool is suddenly depleted.</p>
<h2><a name="On-the-role-of-thermolabile-proteins-in-microalgae" class="anchor" href="#On-the-role-of-thermolabile-proteins-in-microalgae">On the role of thermolabile proteins in microalgae</a></h2>
<p>Mutations drive the evolutionary development. Simultaneously repair and quality check mechanisms prevent harmful mutations in the short term. Are there evolutionary mechanisms that prevent specific proteins from being mutated?
The substitution rate for chloroplastidic genes was calculated to be 3.3*10<sup>-10</sup>, which is almost 10 times lower than in genes of nuclear genomes, probably because of the plastid homologous recombination machinery and the dynamics of chloroplast genomes.
Are there proteins that are intentionally do not undergo stabilizing mutations and thereby are kept thermolabile?</p>
<p>The aim is to identify genes with low evolutionary rates by using the <em>K<sub>a</sub>/K<sub>s</sub> ratio</em> (sometimes referred as d<sub>N</sub>/d<sub>S</sub>, which describes the evolutionary rate of a DNA sequence.</p>
<ul>
<li>K<sub>a</sub> is calculated by dividing all nonsynonymous substitutions per nonsynonymous site, or in other words: the number of sites where an nonsynonymous amino acid substitution occurred per all sites at which a nonsynonymous substitution is possible</li>
<li>K<sub>s</sub> is calculated by dividing all synonymous substitutions per synonymous site</li>
</ul>
<img style="max-width:70%" src="../img/evolution_master.png"></img>
<br>
<img style="max-width:70%" src="../img/evolution_master3.png"></img>
<br>
<p>The ratio <em>k</em> of nonsynonymous (K<sub>a</sub>) to synonymous (K<sub>s</sub>) nucleotide substitution rates is an indicator of selective pressures on genes.</p>
<ul>
<li><p><em>k</em> &gt; 1: more nonsyn than syn mutations -&gt; positive darwinian selection; driving change</p></li>
<li><p><em>k</em> = 1: equal rate of mutations -&gt; neutral evolution</p></li>
<li><p><em>k</em> < 1: more syn than nonsyn mutations -> negative/purifying selection; acting against change</p></li>
</ul>
<p>Candidates can be investigated by comparing mutation rates of all proteins over several species within the <em>Chlorophyta</em> phylum and identify anomalies.
Thereby, proteins found in aggregates are of particular interest since these are thermolabile and hypothetically should be under evolutionary pressure to become more stable.
If there is an equal slow/negative selection of orthologs between species, it’s a clue for thermolability as a mechanism rather than a damage.
Of course, you have to consider, that a strong conservation may be due to preserving the proteins’ function. Accordingly, you have to make sure to keep the proteins affinity/activity
in mind that may require further investigation of mutants proposed by the result of the analysis.</p>
<img style="max-width:70%" src="../img/evolution_master2.png"></img>
<br>
<h1><a name="Project-outline" class="anchor" href="#Project-outline">Project outline</a></h1>
<ul>
<li>
collect various genomes from <em>Chlorophyta</em> phylum
<ul>
<li>various genomes are available upon request</li>
</ul>
</li>
<li>multiple sequence alignment of proteins using ClustalW</li>
<li>Construction of an adequate substitution model since some mutations are far more likely than others.</li>
<li>Check the evolution rate of the proteins across the phylum. If the evolution rate is equal (or equally low) for a specific protein, it may hint to a conserved strategy of the protein</li>
<li>
In previous work we identified transcripts that were up regulated while protein abundance decreased during heat acclimation. This anticorrelation points to specific regulatory role or thermolability.
<ul>
<li>is there an overlap of both analysis?</li>
</ul>
</li>
<li>
How is the ratio of silent/non silent mutations, especially in RubisCO activase?
<ul>
<li>
RubisCO activase aggregates during the first period of heat treatment leading to starch depletion and increased TAG-production due to increased free energy reduction equivalents from photosynthesis.
After the membrane remodelling the amount of RubisCO activase recovers.
</li>
<li><em>Hemme, D., Veyel, D., Mühlhaus, T […] Schroda M. (2014). Systems-Wide Analysis of Acclimation Responses to Long-Term Heat Stress and Recovery in the Photosynthetic Model Organism Chlamydomonas reinhardtii. The Plant cell 26, 4270-4297.</em></li>
</ul>
</li>
</ul>
<h2><a name="Questions-to-answer" class="anchor" href="#Questions-to-answer">Questions to answer</a></h2>
<ul>
<li>Is there a mechanism preserving proteins from becoming more thermostable?</li>
<li>
<p>Where does this constraint come from?</p>
<ul>
<li>Is this conservation due to preserving the enzymatic activity or to preserve thermolability of key regulators?</li>
</ul>
</li>
<li>
<p>What is the benefit?</p>
<ul>
<li>Do affected proteins act as regulators for stress response?</li>
<li>Is there a requirement for a fast depletion? (reduced energy consumption and more free energy for stress responses)</li>
</ul>
</li>
</ul>
<h1><a name="Additional-information" class="anchor" href="#Additional-information">Additional information</a></h1>
<ul>
<li>
Silent mutations are not necessarily 'silent' because altered mRNA sequences may affect transcription, splicing procedures, mRNA transport or translation, even if the protein sequence is identical in the end.
Translation can be affected by differential codon usage.
</li>
<li>A further possibility to preserve a certain kind of thermolability is the requirement for sufficient flexibility to undergo conformational changes during its catalytic activity.</li>
</ul>
<h1><a name="References" class="anchor" href="#References">References</a></h1>
<ul>
<li>How many nucleotide substitutions actually took place?, Jukes TH, 1990</li>
<li>Evolutionary perspectives on protein structure, stability, and functionality, Goldstein Richard; in 'Biological Evolution and Statistical Physics', p 82, Springer 2001</li>
<li>(Introductional) Was ist molekulare Evolution? Die molekulare Uhr; in FJ Ayala, Die großen Fragen Evolution, p 80,  DOI 10.1007/978-3-642-33006-3_10,</li>
<li>(Introductional) 3.12 Molekulare Systematik, Zrzavýin et al.; in Evolution, p 199, Spektrum 2003</li>
<li>Evolutionary Rates and Expression Level in Chlamydomonas, Popescu et al., 2005, 10.1534/genetics.105.047399</li>
<li>(Introductional) Gene, Zufall, Selektion - Populäre Vorstellungen zur Evolution und der Stand des WIssens, Veiko Krauß, Springer, 2014</li>
<li>Codon substitution in evolution and the saturation of synonymoous changes, Gojobori, Genetics Society of America, 1983</li>
<li>POTION: an end-to-end pipeline for positive Darwinian selection detection in genome-scale data through phylogenetic comparison of protein-coding genes, Hongo et al., 2015</li>
<li>Computational Modeling for Evolution of HSP90A Homologues, Datta et al., IJETTCS, 2017</li>
<li>KaKs_Calculator 2.0: A Toolkit Incorporating Gamma-Series Methods and Sliding Window Strategies, Wang, GENOMICS PROTEOMICS &amp; BIOINFORMATICS, 2010</li>
<li>Statistical methods for detecting molecular adaptation, Yang &amp; Bielawski, TREE, 2000</li>
<li>Estimating Synonymous and Nonsynonymous Substitution Rates Under Realistic Evolutionary Models, Yang &amp; Nielsen, 2000</li>
<li><em>Computing Ka and Ks with a consideration of unequal transitional substitutions, Zhang et al., BMC, 2006</em></li>
<li><em>Determinants of the rate of protein sequence evolution, Zhang &amp; Yang, Nature revies Genetics, 2015</em></li>
<li><a href="http://bioinformatics.cvr.ac.uk/blog/calculating-dnds-for-ngs-datasets/">http://bioinformatics.cvr.ac.uk/blog/calculating-dnds-for-ngs-datasets/</a></li>
</ul>

                </div>
            </div>
        </div>
        
    </div>

    <!-- BEGIN SEARCH BOX: this adds support for the search box -->
    <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/JavaScript-autoComplete/1.0.4/auto-complete.css" />
    <script type="text/javascript">var fsdocs_search_baseurl = '/BIO-BTE-10-L-4/'</script>
    <script type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/lunr.js/2.3.8/lunr.min.js"></script>
    <script type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/JavaScript-autoComplete/1.0.4/auto-complete.min.js"></script>
    <script type="text/javascript" src="/BIO-BTE-10-L-4/content/fsdocs-search.js"></script>
    <!-- END SEARCH BOX: this adds support for the search box -->
</body>
<footer>
    <script>
        bulmaCollapsible.attach('.is-collapsible');
    </script>
</footer>
</html>